### PR TITLE
Fixes #24871 - Serve intermediate iPXE script

### PR DESCRIPTION
**Pull request by leewaa**
This adds a new iPXE templated adopted from the wiki on [DHCP setup - unmanaged server](https://projects.theforeman.org/projects/foreman/wiki/Fetch_boot_files_via_http_instead_of_TFTP/23#DHCP-setup-unmanaged-server)

The naming for the script and the settings entry description and labelling is not final and welcomes feedback!

### CODE REVIEW

**Comment by leewaa** [comment]
Also I might just add, this is waiting on #6009 which is also waiting on #5988

File: app/controllers/unattended_controller.rb
Comment (ekohl) [inline app/controllers/unattended_controller.rb]: Should these if statements be nested? i.e.
```ruby
if ipxe_request?
  if !@host
    if params[:bootstrap]
      return render_intermediate_template
    else
      render_default_global_template
      return
    end
  elsif !@host.build?
    render_local_boot_template
    return
  end
end
```

Also makes me wonder why the first render returns something while the others don't.

**Review by ekohl — COMMENTED** [review]
The `require_ssl?` function looks also a bit hard to read. Should it be:
```ruby
def require_ssl?
  preview? ? super : false
end
```

File: app/controllers/unattended_controller.rb
Comment (leewaa) [reply app/controllers/unattended_controller.rb]: I wouldn't nest these. Much harder to read imo.

Good catch on the return. I don't believe what the method returns is important since it is just rendering the template. I guess just syntactical sugar....

**Comment by leewaa** [comment]
@ekohl there are allot of changes shown in here which are not relevant to the PR since I have branched of previous work with is not yet merged. Please wait for further review until #6009 and #5988 have been merged

File: app/controllers/unattended_controller.rb
Comment (ekohl) [reply app/controllers/unattended_controller.rb]: To me the nesting makes it's clearer there is one workflow. If you have single `return render_` lines I think it can be easier to understand the ipxe workflow.

**Comment by ekohl** [comment]
It can be useful to mark the PR with `[WIP]` in the title so other people are aware it's not finalized. Do be aware that it'll skip PR message checking and mark the build as failed because of it. It only reruns the test when you push so remove `[WIP]` before pushing the final version.

**Comment by leewaa** [comment]
Great, thanks for the info!

**Comment by mmoll** [comment]
@leewaa please rebase :)

File: app/controllers/unattended_controller.rb
Comment (leewaa) [reply app/controllers/unattended_controller.rb]: I disagree. Having nested if statements makes it hard to read the code, and it takes longer to understand what is happening.  

Having early returns makes the when and why much much clearer to the reader. Rubocop style guid also advises against nested if statements. https://github.com/rubocop-hq/ruby-style-guide#no-nested-conditionals

Comment (timogoebel) [reply app/controllers/unattended_controller.rb]: I know this is not helpful: I don't like both. Can we extract this further? E.g. a class that finds the appropriate template based on conditions and then just renders that? What do you think?

Comment (timogoebel) [reply app/controllers/unattended_controller.rb]: To make it clearer:

We have a lot of methods like
```ruby
template = ProvisioningTemplate.where(...)
if template
   sefe_render(template)
else
  render_error
end
```

I think we should clean this up and then this will go away.

A service that finds the appropriate template for a request.

Comment (leewaa) [reply app/controllers/unattended_controller.rb]: I think that is getting too extreme. Rendering should be in the controller and not in an external class. 

Due to the complexity of the unattended controller there is no way of avoiding all of the early returns. The controller is responsible for too many things. Sure its all about unattended installations but maybe we can think about splitting this one up into smaller, more focused controllers.... OR we can just accept complexity and keep the early returns and try and keep the code as readable as possible.

However I would stay well clear of nested if statements or extracting out the `render` calls and logic to an external class. 

I agree however that this can be dried up:
```ruby
if template
   sefe_render(template)
else
  render_error
end
```

Comment (timogoebel) [reply app/controllers/unattended_controller.rb]: > Rendering should be in the controller and not in an external class.

I fully agree, but we could create a service that finds the correct template for a request. The controller can then do the rendering. WDYT?

Comment (ekohl) [reply app/controllers/unattended_controller.rb]: Even if it's a helper function within the controller it does make sense to me.

Comment (leewaa) [reply app/controllers/unattended_controller.rb]: This is a step in the right direction. The only hard part is managing the errors that are rendered.

I had a go at cleaning it up but the methods `sefe_render(template)` and `render_custom_error(status, error_message, params)` made it hard. These two should be cleaned up and merged into one but this will require deeper thought and more time.

This is all good discussion but theres one problem. Any changes now will cause us(me) pain with merge conflicts. So lets stay focused on the introduced changes. I would say we focus on  #6009, get it merged. Then the changes here, and then I can look at cleaning up the logic for finding the templates by extracting it out of the controller. sound ok @timogoebel @ekohl  ?

File: app/controllers/unattended_controller.rb
Comment (ekohl) [inline app/controllers/unattended_controller.rb]: I do think this group makes it slightly easier to see but I do think you could at least indent it all with `if ipxe_request?`. Then you can leave out the comment because it's self-documenting. Would that be a fair compromise?

**Comment by lzap** [comment]
And this needs a rebase of course.

**Comment by leewaa** [comment]
This is good to review now

**Comment by leewaa** [comment]
@kamils-iRonin care to take a look ?

**Review by kamils-iRonin — DISMISSED** [review]
looks good to me

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (lzap) [inline app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: Since this is ERB now, can you actually use this:

https://github.com/theforeman/foreman_bootdisk/blob/master/app/views/foreman_bootdisk/host.erb#L17

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (lzap) [inline app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb]: Is this intentional? This always exits, not sure what this means.

**Review by lzap — CHANGES_REQUESTED** [review]
Overall looks good.

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (timogoebel) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb]: This looks like a leftover. The actual template begins below.

Comment (leewaa) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb]: Yep, sorry will fix

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (lzap) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: What I mean, this was just for two NICs, server have often 4-8 or more, so just generate this with ERB up to 32 NICs.

**Comment by leewaa** [comment]
Hey guys, sorry I was off for a few days, attending to this one now.

**Comment by leewaa** [comment]
@lzap can you take another look at the script, I made some adjustments as you suggested.
@timogoebel ok for you also like this ?

**Review by timogoebel — DISMISSED** [review]
Looks good to me.

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (lzap) [inline app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:7]: I think the file needs to start with this "shebang" rather than a blank line.

File: app/controllers/unattended_controller.rb
Comment (lzap) [inline app/controllers/unattended_controller.rb]: I guess you want to use `Setting[:intermediate_ipxe_script]` here.

**Review by lzap — CHANGES_REQUESTED** [review]
And a snapshot test for the template would be great to prevent regressions. Otherwise it looks great.

File: app/controllers/unattended_controller.rb
Comment (leewaa) [reply app/controllers/unattended_controller.rb]: ah good catch, thanks

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (leewaa) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:7]: 👍

**Comment by leewaa** [comment]
Ok, all done @lzap

File: app/models/setting/provisioning.rb
Comment (timogoebel) [inline app/models/setting/provisioning.rb]: Could you please add `{ :collection => Proc.new { Hash[ProvisioningTemplate.unscoped.of_kind(:iPXE).map { |tmpl| [tmpl.name, tmpl.name] }] } }` here similar to the `default_local_boot_templates`? That way a user gets a list of templates to choose from and not a text field.
And could you also add [a validator](https://github.com/theforeman/foreman/blob/5e4268bb583fc665b070ec68e96eb70a152c11da/app/models/setting/provisioning.rb#L15), that this template actually exists (if set via the api)?

Comment (lzap) [reply app/models/setting/provisioning.rb]: Good point let's do that.

**Comment by leewaa** [comment]
Rebased and addressed @timogoebel's comment :)

File: app/controllers/unattended_controller.rb
Comment (lzap) [inline app/controllers/unattended_controller.rb:95]: The name is not available in this context:

```
NameError (undefined local variable or method `name' for #<UnattendedController:0x00005555621517a8>):
  
app/controllers/unattended_controller.rb:94:in `render_intermediate_template'
 | app/controllers/unattended_controller.rb:135:in `render_ipxe_template'
 | app/controllers/unattended_controller.rb:58:in `host_template'
 | app/controllers/concerns/application_shared.rb:15:in `set_timezone'
 | app/models/concerns/foreman/thread_session.rb:32:in `clear_thread'
 | app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller'
 | lib/middleware/telemetry.rb:9:in `call'
 | lib/middleware/catch_json_parse_errors.rb:8:in `call'
 | lib/middleware/logging_context.rb:23:in `call'
  Rendered /home/lzap/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/actionpack-5.2.1/lib/action_dispatch/middleware/templates/rescues/diagnostics.html.erb within rescues/layout (23.7ms)
```

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (lzap) [inline app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: I still think this is worth having a snapshot for. This is code uncovered with tests, at least knowing it renders fine is a great value.

**Review by lzap — CHANGES_REQUESTED** [review]
Would you mind adding a snapshot template test for the script? Also one small error during testing.

File: app/controllers/unattended_controller.rb
Comment (leewaa) [reply app/controllers/unattended_controller.rb:95]: ahh, how did that get there. Copy pasta I guess

File: app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb
Comment (leewaa) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: The controller test that I added does not suffice ?

Comment (leewaa) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: Actually, I see what you mean now, I will get to this.

Comment (lzap) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: Yep, that's why snapshotting is really nice feature, we tend to stub templates with "hello word" :-)

Comment (leewaa) [reply app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb:22]: OK 👍 I will squeeze this in tomorrow.

**Comment by lzap** [comment]
Related test failures.

File: test/controllers/unattended_controller_test.rb
Comment (timogoebel) [inline test/controllers/unattended_controller_test.rb]: This should be `iPXE intermediate` (no `s`) to fix the tests.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt
Comment (timogoebel) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: Newlines are missing here.

File: test/controllers/unattended_controller_test.rb
Comment (leewaa) [reply test/controllers/unattended_controller_test.rb]: yep got this

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt
Comment (leewaa) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: ah, cheers

**Comment by leewaa** [comment]
Ok, I hope there isn't anything else missing 😸

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt
Comment (timogoebel) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: still a missing newline. `chain` needs to be a separate line.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt
Comment (timogoebel) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: Does this really need to be `net31:net31` - the wiki page just has `net31`.
But maybe there is just a newline missing.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt
Comment (leewaa) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: Ah, damnit. So I guess the rake take which generates this isn't working properly.

Comment (leewaa) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: Answering my own assumption: Nope, the template is incorrect, will adjust it

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt
Comment (leewaa) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/iPXE/iPXE intermediate script.snap.txt]: yeah in needs to be there, but not like this... will adjust. FYI for those reading, shoudl be looking like this for each nic:

```
:net1
isset ${net1/mac} || goto no_nic
dhcp net1 || goto net2
chain http://FOREMAN_OR_PROXY/unattended/iPXE?mac=${net1/mac} || goto net2
```

**Comment by lzap** [comment]
I still see a lot of RED. :-(

**Comment by timogoebel** [comment]
And we're 💚.

**Review by lzap — APPROVED** [review]
Looks good to me! Thanks, very nice one.

"32 NICs ought to be enough for everybody." -- lzap, 2018

I am setting RTM, this is good and low risk for 1.20 since this controller was heavily refactored in this cycle (if there are bugs, they are in `develop` already). However, @tbrisker is the one to decide, we are in the silent period.

**Comment by lzap** [comment]
Please file PR into community templates with the same content as here for `app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb`

File: app/models/setting/provisioning.rb
Comment (lzap) [inline app/models/setting/provisioning.rb]: Oh I noticed this too late, but still. The default value does not correspond the template name which is "iPXE intermediate script", therefore bootstrapping always gives 404.

**Review by lzap — CHANGES_REQUESTED** [review]
Realized one more thing.

**Comment by tbrisker** [comment]
Considering we're branching today, and this is a new feature and not a crucial bug fix, i'd rather postpone that to 1.21 if that's ok with you.

File: app/models/setting/provisioning.rb
Comment (leewaa) [reply app/models/setting/provisioning.rb]: Ah! Nice. I will push a fix.

**Comment by leewaa** [comment]
@tbrisker fine by me :)
@lzap, community templates PR is already there, but needs adjusting... will attend to it

File: app/models/setting/provisioning.rb
Comment (lzap) [inline app/models/setting/provisioning.rb]: Cleaning my eyes twice, but I still see the same. Lemme use the GH beta feature now:

```suggestion
      self.set('intermediate_ipxe_script', N_('Intermediate iPXE script for unattended installations'), 'iPXE intermediate script', N_('iPXE intermediate script'), nil, { :collection => Proc.new { Hash[ProvisioningTemplate.unscoped.of_kind(:iPXE).map { |tmpl| [tmpl.name, tmpl.name] }] } }),
```

Comment (leewaa) [reply app/models/setting/provisioning.rb]: oh I see 😂 thanks for the clarification!!

**Review by lzap — APPROVED** [review]
Thanks this is a go, letting @tbrisker to merge it after 1.20 is built successfully.

**Comment by leewaa** [comment]
Great! Cheers for all the input @lzap @ekohl @timogoebel

**Comment by lzap** [comment]
> Please file PR into community templates with the same content as here for `app/views/unattended/provisioning_templates/iPXE/ipxe_intermediate_script.erb`

@leewaa hey I wonder have you filed a PR into community-templates repo with the script? I cannot find it. I can do this myself if you don't intend to do it, just let me know thanks.

**Comment by lzap** [comment]
Found it! https://github.com/theforeman/community-templates/pull/512

**Pull request merged**
